test/topology_experimental_raft/test_tablets.py

```diff
@@ -1222,3 +1240,206 @@ async def test_tablet_storage_freeing(manager: ManagerClient):
+
+    decommision_rack = f"rack{initial_num_racks}"
+    for s in servers.values():
+        if s.rack == decommision_rack:
+            logger.debug(f"Decommissioning server={s}")
+            await manager.decommission_node(s.server_id)
+
+    tablet_count = await get_tablet_count_per_shard_for_hosts(shards_count, manager, list(servers.values()), tables)
+    logger.debug(f"After decommission: expected_replicas_per_shard={expected_replicas_per_shard} tablet_count={tablet_count}")
+    total = 0
+    for server_id, host_replicas in tablet_count.items():
+        s = servers[server_id]
+        for i in host_replicas:
+            total += i
+            if s.rack != decommision_rack:
+                assert abs(i - expected_replicas_per_shard) <= 1
+            else:
+                assert i == 0
+    assert total == initial_tablets * rf
+
+@pytest.mark.asyncio
+async def test_decommission_not_enough_racks(manager: ManagerClient):
+    """
+    Test that decommissioning a rack fails if the number of rack is
+    insufficient to satisfy replication factor, even if the number of
+    nodes is sufficient.
+    Reproduces https://github.com/scylladb/scylladb/issues/19475
+    """
+    logger.info("Bootstrapping multi-rack cluster")
+    servers = dict[ServerNum, ServerInfo]()
+    num_racks = 3
+    nodes_per_rack = 2
+    shards_count = 2
+    initial_tablets = 64
+    rf = num_racks
+    ks = "test"
+    table = "test"
+    for rack in range(1, num_racks + 1):
+        rack_servers = await manager.servers_add(nodes_per_rack, property_file={"dc": "dc1", "rack": f"rack{rack}"})
+        for s in rack_servers:
+            servers[s.server_id] = s
+
+    logger.info("Populate data")
+    cql = manager.get_cql()
+    await cql.run_async(f"CREATE KEYSPACE {ks} WITH replication = {{'class': 'NetworkTopologyStrategy', 'replication_factor': {rf}}} AND tablets = {{'initial': {initial_tablets}}};")
+    await cql.run_async(f"CREATE TABLE {ks}.{table} (pk int PRIMARY KEY, c int);")
+
+    keys = range(100)
+    await asyncio.gather(*[cql.run_async(f"INSERT INTO {ks}.{table} (pk, c) VALUES ({k}, 1);") for k in keys])
+
+    logger.info("Verify tablet replicas distribution")
+    tables = {ks: [table]}
+    expected_replicas_per_shard = (initial_tablets * rf) / (num_racks * nodes_per_rack) / shards_count
+    tablet_count = await get_tablet_count_per_shard_for_hosts(shards_count, manager, list(servers.values()), tables)
+    logger.debug(f"Before decommission: expected_replicas_per_shard={expected_replicas_per_shard} tablet_count={tablet_count}")
+    total = 0
+    for server_id, host_replicas in tablet_count.items():
+        s = servers[server_id]
+        for i in host_replicas:
+            total += i
+            assert abs(i - expected_replicas_per_shard) <= 1
+    assert total == initial_tablets * rf
+
+    decommision_rack = f"rack{num_racks}"
+    decommision_count = 0
+    decommisioned_nodes = set[ServerNum]()
```

Comment:
This is very interesting construct, I have never seen typing used like this. I would expect it to look like:
`decommisioned_nodes: set[ServerNum] = set()`
I could not find this expression in https://peps.python.org/pep-0526/, is something new I missed?

Reply:
I just used it naively as it made sense to me.
I don't mind using the "standard" notation of `decommisioned_nodes: set[ServerNum] = set()`

Reply:
made conservative in next version (though I like the alternative, non-standard notion better BTW)